cmd/localapiclient/localapiclient.go

```diff
@@ -1,87 +1,82 @@
 package localapiclient
 
 import (
-	"bytes"
-	"errors"
+	"context"
 	"fmt"
-	"log"
+	"io"
 	"net/http"
 	"time"
 
 	"tailscale.com/ipn/localapi"
 )
 
-// LocalAPIResponseWriter substitutes for http.ResponseWriter in order to write byte streams directly
-// to a receiver function in the application.
-type LocalApiResponseWriter struct {
-	headers http.Header
-	body    bytes.Buffer
-	status  int
+// Response represents the result of processing an http.Request.
+type Response struct {
+	headers    http.Header
+	status     int
+	bodyWriter io.WriteCloser
+	bodyReader io.ReadCloser
 }
 
-func newLocalApiResponseWriter() *LocalApiResponseWriter {
-	return &LocalApiResponseWriter{headers: http.Header{}, status: http.StatusOK}
-}
-
-func (w *LocalApiResponseWriter) Header() http.Header {
-	return w.headers
+func (r *Response) Header() http.Header {
+	return r.headers
 }
 
 // Write writes the data to the response body, which will be sent to Java. If WriteHeader is not called
 // explicitly, the first call to Write will trigger an implicit WriteHeader(http.StatusOK).
-func (w *LocalApiResponseWriter) Write(data []byte) (int, error) {
-	if w.status == 0 {
-		w.WriteHeader(http.StatusOK)
+func (r *Response) Write(data []byte) (int, error) {
+	if r.status == 0 {
+		r.WriteHeader(http.StatusOK)
 	}
-	return w.body.Write(data)
+	return r.bodyWriter.Write(data)
 }
 
-func (w *LocalApiResponseWriter) WriteHeader(statusCode int) {
-	w.status = statusCode
+func (r *Response) WriteHeader(statusCode int) {
+	r.status = statusCode
 }
 
-func (w *LocalApiResponseWriter) Body() []byte {
-	return w.body.Bytes()
+func (r *Response) Body() io.ReadCloser {
+	return r.bodyReader
 }
 
-func (w *LocalApiResponseWriter) StatusCode() int {
-	return w.status
+func (r *Response) StatusCode() int {
+	return r.status
 }
 
-type LocalApiClient struct {
+type LocalAPIClient struct {
 	h *localapi.Handler
 }
 
-func NewLocalApiClient(h *localapi.Handler) LocalApiClient {
-	return LocalApiClient{h: h}
+func New(h *localapi.Handler) *LocalAPIClient {
+	return &LocalAPIClient{h: h}
 }
 
-var ErrBadHttpStatus = errors.New("bad http status for localapi response")
+// Call calls the given endpoint on the local API using the given HTTP method
+// optionally sending the given body. It returns a Response representing the
+// result of the call and an error if the call could not be completed or the
+// local API returned a status code in the 400 series or greater.
+// Note - Response includes a response body available from the Body method, it
+// is the caller's responsibility to close this.
+func (cl *LocalAPIClient) Call(ctx context.Context, method, endpoint string, body io.Reader) (*Response, error) {
+	ctx, cancel := context.WithTimeout(ctx, 2*time.Second)
```

Comment:
I really like that you included a timeout. A more idiomatic and less verbose way to do this is with contexts.

Note also that passing a context to `Call` is idiomatic and useful if we have some larger event that should cancel processing. In practice, we're passing the appCtx, which means that when that context is cancelled, any in-flight requests will be cancelled too, which is convenient.

Reply:
It looks like the timeout now only applies to making the request - is that right? It'd be nice to have timeouts for the rest of the work done by this function too.

Reply:
Great question! Because we use the function [http.NewRequestWithContext](https://pkg.go.dev/net/http#NewRequestWithContext), the context governs the entire lifetime of the request. As noted in the doc:

> For an outgoing client request, the context controls the entire lifetime of a request and its response: obtaining a connection, sending the request, and reading the response headers and body.

Reply:
Since localapi has its own implementation of ServeHTTP, though, can we expect that this is still true? Iiuc, the handlers in localapi should use r.ctxt?

Reply:
Oh yikes, good point 🤦   So you will need something like your OG implementation of `Call()`, maybe just with the added feature of respecting the deadline on a passed-in `ctx` and returning a body reader that also respects the context?